File: crates/taflow-core/src/stream/cumulative_product.rs
//! Persistent cumulative product state.

use super::StreamingIndicator;
use crate::error::TaResult;

/// Compute the cumulative product of chronological scalar observations.
#[derive(Debug, Clone)]
pub struct CumulativeProduct {
    total: f64,
    value: Option<f64>,
}

impl CumulativeProduct {
    /// Create a fresh cumulative state.
    pub fn new() -> TaResult<Self> {
        Ok(Self::default())
    }

    /// Append one value and return the current cumulative result.
    pub fn append(&mut self, input: f64) -> f64 {
        self.total *= input;
        self.value = Some(self.total);
        self.total
    }

    /// Append a slice into `output` in scalar replay order.
    pub fn extend_slice_into(&mut self, input: &[f64], output: &mut Vec<f64>) {
        output.reserve(input.len());
        output.extend(input.iter().map(|&input| self.append(input)));
    }

    /// Return the latest result, or `None` before the first observation.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Restore fresh-state behavior without reallocating.
    pub fn reset(&mut self) {
        self.total = 1.0;
        self.value = None;
    }
}

impl Default for CumulativeProduct {
    fn default() -> Self {
        Self {
            total: 1.0,
            value: None,
        }
    }
}

impl StreamingIndicator for CumulativeProduct {
    type Output = f64;

    fn append(&mut self, input: f64) -> Option<Self::Output> {
        Some(Self::append(self, input))
    }

    fn value(&self) -> Option<Self::Output> {
        Self::value(self)
    }

    fn reset(&mut self) {
        Self::reset(self);
    }
}

File: python/taflow/indicators/cumulative_product.py
"""Persistent cumulative product indicator."""

from typing import Any

import numpy as np

from .._native import CumulativeProduct as _NativeCumulativeProduct
from .._series import as_float64_series


class CumulativeProduct:
    """Compute the cumulative product in persistent Rust state.

    Supply the chronological numeric ``_input`` series through ``extend``. There is no warm-up. Correctness maps to the
    Polars ``Series.cum_prod`` expression.
    """

    def __init__(self) -> None:
        self._state = _NativeCumulativeProduct()

    def append(self, _input: float) -> "CumulativeProduct":
        """Append one observation and return this indicator."""
        self._state.append(float(_input))
        return self

    def extend(self, _input: Any) -> "CumulativeProduct":
        """Append chronological observations and return this indicator."""
        self._state.extend(as_float64_series(_input))
        return self

    def compute(self) -> np.ndarray:
        """Return the complete aligned ``float64`` history."""
        return self._state.compute()

    @property
    def value(self) -> float | None:
        """Return the latest result, or ``None`` before the first value."""
        return self._state.value

    def reset(self) -> "CumulativeProduct":
        """Restore fresh native state and return this indicator."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of processed observations."""
        return len(self._state)
